### PR TITLE
add docker swarm mode support

### PR DESCRIPTION
**Pull request by pdemagny**
Signed-off-by: Pierre DEMAGNY <pdemagny@printoclock.com>

### CODE REVIEW

**Comment by xcompass** [comment]
Looks good to me. Though I'm never run swarm mode. So I may not be in a good position to review.

File: docker-swarm.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            MM_USERNAME: mmuser
+            MM_PASSWORD: mmuser_password
+            MM_DBNAME: mattermost
+            # in case your config is not in default location
+            # MM_CONFIG=/mattermost/config/config.json
+        deploy:
+            restart_policy:
+                condition: on-failure
+    web:
+        image: nginx:latest
```
Comment (pichouk) [inline docker-swarm.yml]: Why not using the `mattermost/mattermost-prod-web` image ?

File: web/docker-entry.sh [hunk 2]
```diff
@@ -5,5 +5,6 @@ if [ -f "/cert/cert.pem" -a -f "/cert/key-no-password.pem" ]; then
 else
   echo "linking plain config"
 fi
-ln -s /etc/nginx/sites-available/mattermost$ssl /etc/nginx/conf.d/mattermost.conf
+if [ -e "/etc/nginx/conf.d/default.conf" ]; then rm /etc/nginx/conf.d/default.conf; fi
+ln -s /etc/nginx/sites-available/mattermost${ssl} /etc/nginx/conf.d/mattermost.conf
```
Comment (pichouk) [inline web/docker-entry.sh]: Good point here, but is that necessary to add this to this PR ?

To keep a correct Git history I will prefer to remove this changes from this Pull Request and to open a new one just for this (or if you don't want, I'll do the commit by myself). I will accept directly ;)

**Comment by pichouk** [comment]
@pdemagny Thanks for your help !

**Comment by pichouk** [comment]
Also, it seems that `mattermod` bot is not working (@jasonblais ?) and didn't post the comment about Mattermost CLA.
So @pdemagny can you please sign the Mattermost CLA ? More information about this is available on another PR comment https://github.com/mattermost/mattermost-docker/pull/171#issuecomment-329997685

**Comment by jasonblais** [comment]
Thanks @pichouk, posted a note about mattermod in the Developers channel: https://pre-release.mattermost.com/core/pl/8k1wnzk6ppb6pmppuocueqt6kc

**Comment by pdemagny** [comment]
@pichouk Yes ofc i did test it, i wrote it & ran it within a single node swarm with ssl enabled using letsencrypt, tested a few integrations (fabric & jenkins).
I didn't realize there was a mattermost-prod-web image. I'll use it instead of nginx image, which will make the change on docker-entry.sh irrelevant, so i'll remove it. I had other improvements to make anyway, so i'll make the changes later this week or next week.
I'm not a mattermost user/installer normally (which is why i did this, to try it :p), so i'm not familiar with mattermod, but i'll try to look into it too.

**Comment by hmhealey** [comment]
I'm just going to do my best Mattermod impression here since it's not going to pick up this PR now that it's already open.

Thanks for the pull request!

Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

**Comment by pichouk** [comment]
@pdemagny Great ! Ping me when you will need another review :)

For the `mattermod` stuff, it's just a bot that ask you to sign CLA (like @hmhealey does :) ). I can see that you sign it, so it's good now ;)

**Comment by pdemagny** [comment]
actually, ping ;)

**Comment by pdemagny** [comment]
Using official image made my modifications to docker-entry.sh irrelevant as expected.
More importantly, using it also made use of docker configs useless too, which means that this stack file is not tied to docker 17.06+ anymore.

**Comment by pichouk** [comment]
Good for me !

Thanks a lot for this contribution ! 👍 😄

**Pull request merged**